---
title: "Metabolomics analysis of Papaya sensory profiles"
author: "Josh Lomax and Ido Bar"
date: "`r format(Sys.Date(), '%d %B %Y')`"
always_allow_html: yes
output: 
    rmdformats::readthedown:
 #   bookdown::html_document2:
      use_bookdown: TRUE
      includes:
        in_header: style/header.html
        after_body: style/external-links-js.html
      df_print: paged
      # theme: 
      #   version: 5
      #   bootswatch: simplex #sandstone #zephyr # yeti # united
        # primary: "#6CC3D4"
      highlight: tango
      css: "style/style.css"
      # toc: true
      # toc_float: true
      # toc_depth: 4
  #    highlight: pygments
      number_sections: false
      code_folding: show
#      keep_md: true
#bibliography: style/comparative_genomics.bib
bibliography: ["style/Papaya_sensory_metabolomics.bib"] 
csl: style/springer-basic-improved-author-date-with-italic-et-al-period.csl
---

```{r setup, include=FALSE}
devtools::source_gist("7f63547158ecdbacf31b54a58af0d1cc", filename = "util.R")
knitr::opts_chunk$set(list(echo = TRUE, eval=FALSE, message=FALSE))
fav_packages <- c('tidyverse',  
                  "juba/rmdformats")
# pak::pak(fav_packages)

pacman::p_load(char = basename(fav_packages), update=FALSE, install=FALSE)

# add copy to clipboard icon
# klippy::klippy(c('r', 'bash'), position = c('top', 'right'))
```

## Aims

- Genomic characterisation of _Marteilia sydneyi_  
- Identify genes that play a role during infection of _M. sydneyi_ in Sydney Rock Oysters _Saccostrea glomerata_ (QX disease)  

As demonstrated by @colantonioMetabolomicSelectionEnhanced2022, metabolomic analysis of fruit can lead to amazing findings [@liuMultidimensionalAnalysisFlavor2025; @lomaxMultiomicApplicationsUnderstanding2024].

### Objectives

1. Assemble and annotate the _Marteilia sydneyi_ genome  
2. Perform comparative genomics analysis to assess core and novel genes   
3. Identify unique _M. sydneyi_ sequences towards development of molecular assays   


## General overview of the analysis pipeline:

1.  Data pre-processing:  
    a.  Reads basecalling  
    b.  Quality check  
    c.  Adaptor trimming  
    d.  Post-trim quality check  
2.  Mapping reads to host reference genome  
3.  _De Novo_ genome assembly of non-host reads  
4.  Taxonomic classification of contigs  
5.  Removal of additional host and off-target contigs  
6.  Annotation of pathogen genes  
7.  Comparative genimics analyses  
7.  Summary statistics and visualisation  

## Methods

### DNA Extraction
High molecular DNA was extracted using the MonarchÂ® HMW DNA Extraction Kit for Tissue (New England Biolabs, [NEB T3060](https://www.neb.com/en-au/products/t3060-monarch-hmw-dna-extraction-kit-for-tissue#Product%20Information)) from digestive gland tissue of Sydney rock oysters infected with QX disease. The samples were confirmed to be infected with QX (contain _M. sydneyi_) by a PCR assay using the extracted DNA as template, as described by XXX. DNA integrity, purity and quantity were assessed by gel electrophoresis, Nanodrop and Qubit, respectively.    


## General information

This document was last updated at `r Sys.time()` using R Markdown (built with `r R.version.string`). The source code for this website can be found on <https://github.com/JoshLomax/Papaya_sensory_metabolomics>.

Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. It is especially powerful at authoring documents and reports which include code and can execute code and use the results in the output. For more details on using R Markdown see <http://rmarkdown.rstudio.com>, [R Markdown: The Definitive Guide](https://bookdown.org/yihui/rmarkdown/) and [Rmarkdown cheatsheet](https://rstudio.github.io/cheatsheets/html/rmarkdown.html).

------------------------------------------------------------------------

## Bibliography